tests/common/include/tests/common/path_correction.inl

```diff
@@ -0,0 +1,524 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Project include(s).
+#include "detray/definitions/units.hpp"
+#include "detray/materials/predefined_materials.hpp"
+#include "detray/propagator/actor_chain.hpp"
+#include "detray/propagator/actors/parameter_resetter.hpp"
+#include "detray/propagator/actors/parameter_transporter.hpp"
+#include "detray/propagator/line_stepper.hpp"
+#include "detray/propagator/navigator.hpp"
+#include "detray/propagator/propagator.hpp"
+#include "detray/propagator/rk_stepper.hpp"
+#include "detray/tracks/tracks.hpp"
+#include "tests/common/tools/detector_metadata.hpp"
+
+// Covfie include(s).
+#include <covfie/core/field.hpp>
+#include <covfie/core/field_view.hpp>
+
+// Vecmem include(s).
+#include <vecmem/memory/host_memory_resource.hpp>
+
+// google-test include(s).
+#include <gtest/gtest.h>
+
+using namespace detray;
+
+struct surface_targeter : actor {
+
+    struct state {
+        scalar _path;
+        dindex _target_surface_index = dindex_invalid;
+    };
+
+    /// Enforces thepath limit on a stepper state
+    ///
+    /// @param abrt_state contains the path limit
+    /// @param prop_state state of the propagation
+    template <typename propagator_state_t>
+    DETRAY_HOST_DEVICE void operator()(state &actor_state,
+                                       propagator_state_t &prop_state) const {
+
+        prop_state._heartbeat = true;
+        auto &navigation = prop_state._navigation;
+        auto &stepping = prop_state._stepping;
+        navigation.set_full_trust();
+
+        scalar residual = actor_state._path - stepping.path_length();
+        stepping.set_constraint(residual);
+
+        typename propagator_state_t::navigator_state_type::intersection_t is;
+        is.index = actor_state._target_surface_index;
+        is.path = residual;
+        auto &candidates = navigation.candidates();
+        candidates.clear();
+        candidates.push_back(is);
+        navigation.set_next(candidates.begin());
+        navigation.set_unknown();
+
+        if (residual < std::abs(1e-6)) {
+            prop_state._heartbeat = false;
+            candidates.push_back({});
+            navigation.next()++;
+            navigation.set_on_module();
```

Comment:
In principle at least that navigator should figure this out on its own after you manipulate the current candidate. Can you explain what the issue is?

Reply:
I reset the candidate vector every actor call (`candidates.clear()`) so I have to set the status manually. Now you might wonder why I am resetting the vector and push the element (`candidates.push_back(is)`): In the detector and track setup I made in `path_correction.inl` where the track should make a full turn starting from a surface, the navigation would abort with the original navigation algorithm because it cannot find any candidate surface.

I expected that you would not like the actor but I thought it is better than making a new navigator.

Of course, the most desirable way is enhancing the existing navigator algorithm so that we can use it without any redundant actor class.

Reply:
What I am afraid of is, that adding an unsafe method to a class interface inevitably leads to people using it also in other contexts, which can lead to weird side effects and bugs. So, yes, you are right unfortunately, I am not a big fan of this change :p

Reply:
In that sense, isn't `set_state`also pretty dangerous?

Reply:
But otherwise, you are right, of course, I would consider that dangerous, too

Reply:
There is one function that is even more dangerous, that I am also not happy about and that is ```set_volume```. I just haven't figured out a better way yet. Probably we should use the state constructor to set the start volume of the propagation?